src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+
+    JSONObject jsonObject = new JSONObject(jsonResponseString);
+
+    if (!jsonObject.keySet().contains(teamProjectRole)) {
+      logger.warn("User is not authorized to access this team project's data");
+      return false;
+    } else {
+      JSONArray teamProjectAuthorizations = jsonObject.getJSONArray(teamProjectRole);
+      logger.debug("Found authorizations={}", teamProjectAuthorizations);
+      // We expect only one authorization rule per teamproject:
+      if (teamProjectAuthorizations.length() != 1) {
+        logger.error("Only one authorization rule expected for 'teamproject'={}, found={}", teamProjectRole,
+          teamProjectAuthorizations.length());
+        return false;
+      }
+      JSONObject teamProjectAuthorization = teamProjectAuthorizations.getJSONObject(0);
+
+      // check if the authorization contains the right "service" and "method" values:
+      String expectedMethod = "access";
+      String expectedService = "atlas-argo-wrapper-and-cohort-middleware"; // TODO - make the service name configurable?
+      String service = teamProjectAuthorization.getString("service");
+      String method = teamProjectAuthorization.getString("method");
+      logger.debug("Parsed service={} and method={}", service, method);
+      if (!method.equalsIgnoreCase(expectedMethod)) {
+        logger.error("The 'teamproject' authorization method should be '{}', but found '{}'", expectedMethod, method);
+        return false;
+      }
+      logger.debug("Parsed method is as expected");
+      if (!service.equalsIgnoreCase(expectedService)) {
+        logger.error("The 'teamproject' authorization service should be '{}', but found '{}'", expectedService, service);
+        return false;
+      }
+      logger.debug("Parsed service is as expected");
+      return true;
+    }
+  }
+
+  /**
+   * Tries to find the team project information in the given request or as part of the
+   * stored this.authorizer.getCurrentTeamProjectRoleForCurrentUser(). Returns null
+   * if nothing can be found.
+   */
+  private String extractTeamProjectFromRequestParameters(ServletRequest request) throws Exception {
+    // Get the url
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+    String url = httpRequest.getRequestURL().toString();
+
+    RoleEntity currentTeamProjectRole = this.authorizer.getCurrentTeamProjectRoleForCurrentUser();
+    String currentTeamProjectName = null; 
+    if (currentTeamProjectRole != null) {
+      currentTeamProjectName = this.authorizer.getCurrentTeamProjectRoleForCurrentUser().getName();
+      if (currentTeamProjectName == null || currentTeamProjectName.trim().isEmpty()) {
+        throw new Exception("The teamproject role was found but name was unexpectedly empty");
+      }
+    }
+    logger.debug("Current teamproject: {}...", currentTeamProjectName);
+    logger.debug("Checking if a teamproject has been specified in the request...");
```

Comment:
Is this case needed? If not, we should probably remove it for the security reasons mentioned above.

Reply:
see previous comments regarding the authorization check.

Reply:
I understand the authorization check but that would come after the GET had revealed the team. A user who successfully logs in might find the URL, copy it with the team, and paste it into a document that becomes public. This reveals the ID of the team project to others. That information reveal probably is not necessary and creates a small risk of hacking through methods such as perhaps social hacking.

Reply:
See also my previous reply on the concern regarding exposing the team project code. Furthermore, we are actually showing the team project code on the UI as well, so currently we are not trying to hide it at all.

Reply:
marking resolved, as discussed today.